xtask/src/release.rs

```diff
@@ -0,0 +1,162 @@
+
+        cmd!("git push {remote} {tag}").run()?;
+
+        let request_body = &json!({
+            "tag_name": tag,
+            "name": name,
+            "body": changes.trim_softbreaks(),
+        })
+        .to_string();
+
+        cmd!(
+            "curl -u {credentials} -X POST -H 'Accept: application/vnd.github.v3+json'
+            {GITHUB_API_RELEASES} -d {request_body}"
+        )
+        .secret(true)
+        .run()?;
+
+        Ok(())
+    }
+
+    /// Get the changes of the given version from the changelog.
+    fn get_changes(&self, project_root: &Path, version: &str) -> Result<String> {
+        let changelog = read_file(project_root.join(&self.name).join("CHANGELOG.md"))?;
+        let lines_nb = changelog.lines().count();
+        let mut lines = changelog.lines();
+
+        let start = match lines.position(|l| l.starts_with(&format!("# {}", version))) {
+            Some(p) => p + 1,
+            None => {
+                return Err("Could not find version title in changelog".into());
+            }
+        };
+
+        let length = match lines.position(|l| l.starts_with("# ")) {
+            Some(p) => p,
+            None => lines_nb,
+        };
+
+        let changes = changelog.lines().skip(start).take(length).join("\n");
+
+        Ok(changes.trim().to_owned())
+    }
+
+    /// Load the GitHub config from the config file.
+    fn get_remote(&self) -> Result<String> {
+        let branch = cmd!("git rev-parse --abbrev-ref HEAD").read()?;
+        let remote = cmd!("git config branch.{branch}.remote").read()?;
+
+        if remote.is_empty() {
+            return Err("Could not get current git remote".into());
+        }
+
+        Ok(remote)
+    }
+
+    /// Get the current version of the crate from the manifest.
+    fn get_version(&self, project_root: &Path) -> Result<String> {
+        let manifest_toml = read_file(project_root.join(&self.name).join("Cargo.toml"))?;
+        let manifest: CargoManifest = from_toml_str(&manifest_toml)?;
+
+        Ok(manifest.package.version)
+    }
+}
+
+/// The required cargo manifest data of a crate.
+#[derive(Debug, Deserialize)]
+struct CargoManifest {
+    /// The package information.
+    package: CargoPackage,
+}
+
+/// The required package information from a crate's cargo manifest.
+#[derive(Debug, Deserialize)]
+struct CargoPackage {
+    /// The package version.
+    version: String,
+}
+
+/// String manipulations for crate release.
+trait StrExt {
+    /// Remove soft line breaks as defined in CommonMark spec.
+    fn trim_softbreaks(&self) -> String;
+}
+
+impl StrExt for str {
+    fn trim_softbreaks(&self) -> String {
+        let mut string = "".to_owned();
+        let mut current = 0;
+        let mut chars = self.char_indices();
+
+        while let Some(p) = self[current..].find('\n') {
+            let pos = current + p;
+            string.push_str(&self[current..pos]);
+
+            if self[pos..].starts_with("\n\n") {
```

Comment:
`self[pos..]` is used a lot, might be more readable if it is bound to a local variable.